src/pipeline.cpp

```diff
@@ -550,7 +550,7 @@ PJ *OPERATION(pipeline,0) {
     /* determine if an inverse operation is possible */
     for( auto& step: pipeline->steps) {
         PJ *Q = step.pj;
-        if ( pj_has_inverse(Q) ) {
+        if ( step.omit_inv || pj_has_inverse(Q) ) {
```

Comment:
I'm wondering if that shouldn't be:
```suggestion
        if ( (!Q->inverted && step.omit_inv) || (Q->inverted && step.omit_fwd) || pj_has_inverse(Q) ) {
```
but I'm not 100% positive as I don't recall whet "+proj=something +inv +omit_inv" is actually supposed to do...

Reply:
> "+proj=something +inv +omit_inv" is actually supposed to do...

See https://proj.org/operations/pipeline.html#cmdoption-arg-omit_fwd. Basically, in certain circumstances it can be useful to skip a step in a pipeline, for instance if you need to treat the forward and inverse differenctly as explained in the docs

Reply:
> See https://proj.org/operations/pipeline.html#cmdoption-arg-omit_fwd

yes I know. My unknown was on the double negation aspect +inv +omit_inv, which isn't covered in the example, and that in the forward and inverse paths...

So experimenting with:
$ echo 0 0 0 | bin/cct  +proj=pipeline +step +inv +proj=geogoffset +dlon=3600 +omit_inv
 -1.0000000000    0.0000000000        0.0000           inf

it means that in the forward direction, the inverse step is done (and thus +omit_inv without effect).

and when adding -I (triple negation):

$ echo 0 0 0 | bin/cct -I +proj=pipeline +step +inv +proj=geogoffset +dlon=3600 +omit_inv
  0.0000000000    0.0000000000        0.0000           inf

so +omit_inv specific effect isn't affected by the presence or absence of the +inv flag of the step.

which *probably* means that my above suggested change is not correct (my head is still hurting)

Reply:
Here's a table of the combinations

| flags | invertible step fwd | invertible step inv | non-invertible step fwd | non-invertible step inv |
| --- | --- | --- | --- | --- |
| +omit_fwd | omit | inv | omit | undefined |
| +omit_fwd +inv | omit | fwd | omit | fwd |
| +omit_inv | fwd | omit | fwd | **!** (should be omit) |
| +omit_inv +inv | inv | omit | undefined | omit |
| | fwd | inv | fwd | undefined |
| +inv | inv | fwd | undefined | fwd |

The one bold exclamation mark there is the one that's not currently working and what this patch is trying to solve. The step doesn't have a valid inverse, but the inverse path is omitted anyway so it should be OK. Currently the pipeline code marks the whole pipeline as not invertible because of this.

The undefined cells are operations that would try to run the inverse operation when the inverse isn't defined.

I think I can use this table to write gie tests for each of the table cells.

Reply:
> I think I can use this table to write gie tests for each of the table cells.

That would certianly be helpful. Especially if you add comments that explain what's going on. All these negations can be very difficult to wrap your head around.

Reply:
I've added more tests based on the table and I tried to add comments describing what's being tested. I agree that it's very hard to think about multiple negations.